src/Data/Fin/Relation/Unary/Top.agda

```diff
@@ -0,0 +1,135 @@
+-- The View, considered as a unary relation on Fin (suc n)
+
+-- NB `Data.Fin.Properties.fromℕ≢inject₁` establishes that the following
+-- inductively defined family on `Fin (suc n)` has constructors which
+-- target *disjoint* instances of the family; and hence the interpretations
+-- of the View constructors will also be disjoint
+
+data View : (i : Fin (suc n)) → Set where
+
+  top :                View (fromℕ n)
+  inj : (j : Fin n) → View (inject₁ j)
+
+-- The view covering function, witnessing soundness of the view
+
+view : ∀ {n} i → View {n} i
+view {n = zero}  zero    = top
+view {n = suc _} zero    = inj _
+view {n = suc n} (suc i) with view {n} i
+... | top   = top
+... | inj j = inj (suc j)
+
+-- Interpretation of the view constructors
+
+⟦_⟧ : ∀ {i} → View {n} i → Fin (suc n)
+⟦ top ⟧   = fromℕ _
+⟦ inj j ⟧ = inject₁ j
+
+-- Completeness of the view
+
+view-complete : ∀ {i} (v : View {n} i) → ⟦ v ⟧ ≡ i
+view-complete top     = refl
+view-complete (inj j) = refl
+
+-- 'Computational' behaviour of the covering function
+
+view-top : ∀ n → view {n} (fromℕ n) ≡ top
+view-top zero    = refl
+view-top (suc n) rewrite view-top n = refl
+
+view-inj : ∀ j → view {n} (inject₁ j) ≡ inj j
+view-inj zero       = refl
+view-inj (suc j) rewrite view-inj j = refl
+
+------------------------------------------------------------------------
+-- Experimental
+--
+-- Because we work --without-K, Agda's unifier will complain about
+-- attempts to match `refl` against hypotheses of the form
+--    `view {n] i ≡ v`
+-- or gets stuck trying to solve unification problems of the form
+--    (inferred index ≟ expected index)
+-- even when these problems are *provably* solvable.
+--
+-- So the two predicates on values of the view defined below
+-- are extensionally equivalent to the assertions
+-- `view {n] i ≡ top` and `view {n] i ≡ inj j`
+--
+-- But such assertions can only ever have a unique (irrelevant) proof
+-- so we introduce instances to witness them, themselves given in
+-- terms of the functions `view-top` and `view-inj` defined above
+
+module Instances {n} where
+
+  private
+
+    lemma : n ≡ toℕ (inject₁ (fromℕ n))
+    lemma = sym (begin
+      toℕ (inject₁ (fromℕ n)) ≡⟨ toℕ-inject₁ (fromℕ n) ⟩
+      toℕ (fromℕ n)           ≡⟨ toℕ-fromℕ n ⟩
+      n                         ∎) where open ≡-Reasoning
+
+  data IsTop : ∀ {i} → View {n} i → Set where
+
+    top : IsTop top
+
+  instance
+
+    top⁺ : IsTop {i = fromℕ n} (view (fromℕ n))
+    top⁺ rewrite view-top n = top
+
+  data IsInj : ∀ {i} → View {n} i → Set where
+
+    inj : ∀ j → IsInj (inj j)
+
+  instance
+
+    inj⁺ : ∀ {j} → IsInj (view (inject₁ j))
+    inj⁺ {j} rewrite view-inj j = inj j
+
+    inject₁≡⁺ : ∀ {i} {j} {eq : inject₁ i ≡ j} → IsInj (view j)
+    inject₁≡⁺ {eq = refl} = inj⁺
+
+    inject₁≢n⁺ : ∀ {i} {n≢i : n ≢ toℕ (inject₁ i)} → IsInj (view {n} i)
+    inject₁≢n⁺ {i} {n≢i} with view i
+    ... | top with () ← n≢i lemma
```

Comment:
Done, and that lets me inline the lemma as a `where` clause, instead of the `private` block. 

[Comments I had added earlier didn't seem to have propagated properly while connection to github was so bad... I'm tethered to my phone now for data]

Separately, though, I disagree with library convention; in particular, I think the use of irrefutable `with () ← ...`, as with other uses of irrefutable `with`, is a *very* convenient idiomatic usage. Here, it was rendered horrible by not being able to scope the `where` block over the `with`... a separate Agda issue [#5883](https://github.com/agda/agda/issues/5883) that I have raised before.